pkg/extensions/sync/sync.go

```diff
@@ -77,85 +87,148 @@ type Destination interface {
+	return &SyncRepoTaskGenerator{
+		Service:      service,
+		status:       pending,
+		waitTime:     0,
+		lastTaskTime: time.Now(),
+		lock:         &sync.Mutex{},
+		lastRepo:     "",
+		log:          log,
 	}
 }
 
-func (gen *TaskGenerator) Name() string {
+func (gen *SyncRepoTaskGenerator) Name() string {
 	return "SyncGenerator"
 }
 
-func (gen *TaskGenerator) Next() (scheduler.Task, error) {
+func (gen *SyncRepoTaskGenerator) Next() (scheduler.Task, error) {
+	var newTask scheduler.Task
+
+	gen.lock.Lock()
+	defer gen.lock.Unlock()
+
+	if gen.status != pending || time.Since(gen.lastTaskTime) <= gen.waitTime {
+		return newTask, nil
+	}
+
+	gen.status = running
+
 	if err := gen.Service.SetNextAvailableURL(); err != nil {
+		onGeneratorError(gen)
+
 		return nil, err
 	}
 
 	repo, err := gen.Service.GetNextRepo(gen.lastRepo)
 	if err != nil {
+		onGeneratorError(gen)
+
 		return nil, err
 	}
 
 	if repo == "" {
-		gen.log.Info().Str("component", "sync").Msg("finished syncing all repos")
-		gen.done = true
-
-		return nil, nil
-	}
+		gen.log.Info().Str("component", "sync").Msg("finished syncing all repositories")
+		gen.status = done
 
-	// a task with this repo is already running
-	if gen.lastRepo == repo {
 		return nil, nil
 	}
 
 	gen.lastRepo = repo
 
-	return newSyncRepoTask(gen.lastRepo, gen.Service), nil
+	return newSyncRepoTask(gen.lastRepo, gen, gen.Service), nil
 }
 
-func (gen *TaskGenerator) IsDone() bool {
-	return gen.done
+func (gen *SyncRepoTaskGenerator) IsDone() bool {
+	gen.lock.Lock()
+	status := gen.status
+	gen.lock.Unlock()
+
+	return status == done
 }
 
-func (gen *TaskGenerator) IsReady() bool {
+func (gen *SyncRepoTaskGenerator) IsReady() bool {
 	return true
 }
 
-func (gen *TaskGenerator) Reset() {
+func (gen *SyncRepoTaskGenerator) Reset() {
+	gen.lock.Lock()
+	defer gen.lock.Unlock()
+
 	gen.lastRepo = ""
 	gen.Service.ResetCatalog()
-	gen.done = false
+	gen.status = pending
+	gen.waitTime = 0
 }
 
 type syncRepoTask struct {
-	repo    string
-	service Service
+	repo      string
+	service   Service
+	generator *SyncRepoTaskGenerator
 }
 
-func newSyncRepoTask(repo string, service Service) *syncRepoTask {
-	return &syncRepoTask{repo, service}
+func newSyncRepoTask(repo string, generator *SyncRepoTaskGenerator, service Service) *syncRepoTask {
+	return &syncRepoTask{repo, service, generator}
 }
 
 func (srt *syncRepoTask) DoWork(ctx context.Context) error {
-	return srt.service.SyncRepo(ctx, srt.repo)
+	err := srt.service.SyncRepo(ctx, srt.repo)
+	if err != nil {
+		onTaskError(srt.generator)
+
+		return err
+	}
+
+	onTaskSuccess(srt.generator)
+
+	return nil
 }
 
 func (srt *syncRepoTask) String() string {
-	return fmt.Sprintf("{Name: \"%s\", repo: \"%s\"}",
+	return fmt.Sprintf("{Name: \"%s\", repository: \"%s\"}",
 		srt.Name(), srt.repo)
 }
 
 func (srt *syncRepoTask) Name() string {
 	return "SyncTask"
 }
+
+func onGeneratorError(gen *SyncRepoTaskGenerator) {
```

Comment:
Why not have this as a generator method?
Same applies for the other methods below.
Modifying the internal state of the generator object using methods outside of the struct is a bad practice.